src/Formats/EscapingRuleUtils.cpp

```diff
@@ -255,7 +259,212 @@ String readStringByEscapingRule(ReadBuffer & buf, FormatSettings::EscapingRule e
+        }
+
+        /// If we have date/datetimes and smth else, convert them to string.
+        /// If we have only dates and datetimes, convert dates to datetime.
+        if (settings.try_infer_dates || settings.try_infer_datetimes)
+        {
+            bool have_dates = false;
+            bool have_datetimes = false;
+            bool all_dates_or_datetimes = true;
+
+            for (const auto & type : data_types)
+            {
+                have_dates |= isDate(type);
+                have_datetimes |= isDateTime64(type);
+                all_dates_or_datetimes &= isDate(type) || isDateTime64(type);
+            }
+
+            if (!all_dates_or_datetimes && (have_dates || have_datetimes))
+            {
+                for (auto & type : data_types)
+                {
+                    if (isDate(type) || isDateTime64(type))
+                        type = std::make_shared<DataTypeString>();
+                }
+            }
+            else if (have_dates && have_datetimes)
+            {
+                for (auto & type : data_types)
+                {
+                    if (isDate(type))
+                        type = std::make_shared<DataTypeDateTime64>(9);
+                }
+            }
+        }
+
+        if (!is_json)
+            return;
+
+        /// Check settings specific for JSON formats.
+
+        /// If we have numbers and strings, convert numbers to strings.
+        if (settings.json.try_infer_numbers_from_strings)
+        {
+            bool have_strings = false;
+            bool have_numbers = false;
+            for (const auto & type : data_types)
+            {
+                have_strings |= isString(type);
+                have_numbers |= isNumber(type);
+            }
+
+            if (have_strings && have_numbers)
+            {
+                for (auto & type : data_types)
+                {
+                    if (isNumber(type) && (!numbers_parsed_from_json_strings || numbers_parsed_from_json_strings->contains(type.get())))
+                        type = std::make_shared<DataTypeString>();
+                }
+            }
+        }
+
+        if (settings.json.read_bools_as_numbers)
+        {
+            bool have_floats = false;
+            bool have_integers = false;
+            bool have_bools = false;
+            for (const auto & type : data_types)
+            {
+                have_floats |= isFloat(type);
+                have_integers |= isInteger(type) && !isBool(type);
+                have_bools |= isBool(type);
+            }
+
+            if (have_bools && (have_integers || have_floats))
+            {
+                for (auto & type : data_types)
+                {
+                    if (isBool(type))
+                    {
+                        if (have_integers)
+                            type = std::make_shared<DataTypeInt64>();
+                        else
+                            type = std::make_shared<DataTypeFloat64>();
```

Comment:
We cannot have both `have_integers` and `have_floats` equal to true, because in one of the previous checks we convert all integers to floats if we have both. I will write a comment